File: src/BuildingBlocks/NotificationHub.Application/Abstractions/Paging.cs
namespace NotificationHub.Application.Abstractions;

/// <summary>
/// Canonical server-side table query parameters.
/// page is 1-based. sort is a field name; order is asc|desc.
/// </summary>
public sealed record PagedRequest(
    int Page = 1,
    int PageSize = 20,
    string? Sort = null,
    string? Order = "asc",
    string? Search = null)
{
    public int SafePage => Page < 1 ? 1 : Page;
    public int SafePageSize => PageSize < 1 ? 20 : PageSize > 100 ? 100 : PageSize;
    public int Skip => (SafePage - 1) * SafePageSize;
    public bool Descending => string.Equals(Order, "desc", StringComparison.OrdinalIgnoreCase);
}

/// <summary>
/// Canonical server-side table response envelope.
/// </summary>
public sealed record PagedResult<T>(
    IReadOnlyList<T> Items,
    int Page,
    int PageSize,
    int TotalCount)
{
    public int TotalPages => PageSize <= 0 ? 0 : (int)Math.Ceiling(TotalCount / (double)PageSize);
    public bool HasNext => Page * PageSize < TotalCount;
    public bool HasPrevious => Page > 1;

    public static PagedResult<T> Empty(int page = 1, int pageSize = 20) =>
        new([], page, pageSize, 0);
}

File: apps/admin/types/paging.ts
/** Canonical server-side table query + response (mirrors backend PagedRequest / PagedResult). */

export type SortOrder = 'asc' | 'desc'

export interface PagedRequest {
  page?: number
  pageSize?: number
  sort?: string
  order?: SortOrder
  search?: string
}

export interface PagedResult<T> {
  items: T[]
  page: number
  pageSize: number
  totalCount: number
  totalPages?: number
  hasNext?: boolean
  hasPrevious?: boolean
}

export function normalizePagedResult<T>(raw: unknown): PagedResult<T> {
  if (!raw || typeof raw !== 'object') {
    return { items: [], page: 1, pageSize: 20, totalCount: 0, totalPages: 0 }
  }
  const r = raw as Record<string, unknown>
  const items = (r.items ?? r.Items ?? []) as T[]
  const page = Number(r.page ?? r.Page ?? 1)
  const pageSize = Number(r.pageSize ?? r.PageSize ?? 20)
  const totalCount = Number(r.totalCount ?? r.TotalCount ?? 0)
  const totalPages =
    r.totalPages != null
      ? Number(r.totalPages)
      : r.TotalPages != null
        ? Number(r.TotalPages)
        : pageSize > 0
          ? Math.ceil(totalCount / pageSize)
          : 0
  return {
    items: Array.isArray(items) ? items : [],
    page,
    pageSize,
    totalCount,
    totalPages,
    hasNext: Boolean(r.hasNext ?? r.HasNext ?? page * pageSize < totalCount),
    hasPrevious: Boolean(r.hasPrevious ?? r.HasPrevious ?? page > 1),
  }
}
